contracts/6/forwarder/ERC20FeeProxy.sol

```diff
@@ -4,86 +4,76 @@ pragma experimental ABIEncoderV2;
+        return nonce;
+    }
 
-    function erc20Execute(
-        ForwardRequest memory req,
+    function executeEIP712(
+        ERC20ForwardRequest memory req,
         bytes32 domainSeparator,
-        bytes32 requestTypeHash,
-        bytes calldata suffixData,
         bytes calldata sig
         ) 
         external payable 
         returns (bool success, bytes memory ret){
             uint256 initialGas = gasleft();
-            (success,ret) = forwarder.execute.value(req.value)(req,domainSeparator,requestTypeHash,suffixData,sig);
+            (success,ret) = forwarder.executeEIP712.value(req.msgValue)(req,domainSeparator,sig);
+            if ( address(this).balance>0 ) {
+                //can't fail: req.from signed (off-chain) the request, so it must be an EOA...
+                payable(req.from).transfer(address(this).balance);
+            }
             uint256 postGas = gasleft();
-            _transferHandler(req,suffixData,initialGas.sub(postGas));
+            _transferHandler(req,initialGas.sub(postGas));
     }
 
-    function erc20ExecutePersonalSign(
-        ForwardRequest memory req,
-        bytes32 domainSeparator,
-        bytes32 requestTypeHash,
-        bytes calldata suffixData,
+
+    function executePersonalSign(
+        ERC20ForwardRequest memory req,
         bytes calldata sig
         ) 
         external payable 
         returns (bool success, bytes memory ret){
             uint256 initialGas = gasleft();
-            (success,ret) = forwarder.executePersonalSign.value(req.value)(req,domainSeparator,requestTypeHash,suffixData,sig);
+            (success,ret) = forwarder.executePersonalSign.value(req.msgValue)(req,sig);
+            if ( address(this).balance>0 ) {
+                //can't fail: req.from signed (off-chain) the request, so it must be an EOA...
+                payable(req.from).transfer(address(this).balance);
+            }
             uint256 postGas = gasleft();
-            _transferHandler(req,suffixData,initialGas.sub(postGas));
+            _transferHandler(req,initialGas.sub(postGas));
     }
 
-
-    function _transferHandler(ForwardRequest memory req,bytes memory suffixData,uint256 executionGas) internal{
-        (address token,address feeReceiver,address feeMultiplierManager,uint256 price,uint256 gasToCover) = 
-        abi.decode(suffixData,(address,address,address,uint256,uint256));
-        gasToCover = executionGas < gasToCover ? executionGas : gasToCover;
-        uint16 multiplierBasisPoints = IFeeMultiplier(feeMultiplierManager).getFeeMultiplier(req.from,token);
-        require(IERC20(token).transferFrom(req.from,feeReceiver,price.mul(gasToCover.add(transferHandlerGas)).
-        mul(multiplierBasisPoints).
-        div(10000)));
+    //good
+    function _transferHandler(ERC20ForwardRequest memory req,uint256 executionGas) internal{
+        uint16 multiplierBasisPoints = IFeeMultiplier(req.feeMultiplierManager).getFeeMultiplier(req.from,req.token);
```

Comment:
So, if I'm correct in my understanding : we'd essentially have one ERC20FeeProxy per client?

Reply:
With a set domain separator, as well. All ERC20 Fee Proxies would interact with the same Biconomy Trusted Forwarder?

Reply:
So, in terms of architecture, which is better and why:
- Telling clients to inherit ERC20FeeProxy and add the fee multiplier logic themselves (ERC20FeeProxy becomes an abstract contract)
- Deploy ERC20FeeProxy, set feeMultiplier, feeReceiver, permitted tokens etc post deployment (as you suggested)

Reply:
So second approach is better, because of following reason:
1. On client contract, we just want him to inherit the BaseRelayRecipeint only.
2. FeeMultiplier and feeReceiver address should be on contract, because if it is in request, user can change these params before signing on client side only and we'll have to keep a check for these on our backend systems, which'll be extra things to do.
3. Ofcourse we'll provide the existing deployed ERC20 proxy with our receiver address and feeMultiplier implementation with stable tokens supported.
4. If client wants his address as receiver and want to support different tokens then he can just deploy these contracts set the configuration by himself and provide the address on our dashboard and we'll just use that. Thats takes away our burden of having custom checks at our backend.